### PR TITLE
Test generated visitor methods

### PR DESCRIPTION
**Pull request by jdevera**
# 📖 What does this PR do?

TL;DR: It checks that the generated Mypy AST visitor methods are correctly named and typed.

After the conversation in abandoned PR #38 where I attempted to generate the visitor from Mypy directly, it became clear that the levels of hackery required to do so were beyond what we could consider stable for a tool that is meant to run on very different boxes.

However, such approaches are not so dangerous for testing since tests will signal to everybody immediately that they have failed.

So in this new PR, I have recycled a lot of the logic I used to generate the visitor methods but this time I use it to check, in tests, that the current method to node type mapping is exactly the one found on Mypy.

This, funnily enough, revealed there was a spurious method in the list.

### CODE REVIEW

File: test/mypy_visitor.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    def loaded_mypy_modules() -> Iterator[str]:
+        """Covenient block to get names of imported mypy modules"""
+        for mod_name in sys.modules:
+            if mod_name == "mypy" or mod_name.startswith("mypy."):
+                yield mod_name
+
+    # First, backup all imported mypy modules and remove them from sys.modules,
+    # so they will not be found in resolution
+    saved_mypy = dict()
```
Comment (dosisod) [inline test/mypy_visitor.py:108]: Replace `dict()` with `{}`. This is actually a check in Refurb (FURB112), but since Refurb only checks the `refurb` directory, no warning was emitted.

File: test/mypy_visitor.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    for mod_name in list(loaded_mypy_modules()):
+        saved_mypy[mod_name] = sys.modules[mod_name]
+        del sys.modules[mod_name]
```
Comment (dosisod) [inline test/mypy_visitor.py:111]: This could be written like this:

```python
saved_mypy[mod_name] = sys.modules.pop(mod_name)
```

The "no del" check (FURB131) doesn't currently catch this because of the `.` part.

File: test/mypy_visitor.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    clazz.
+
+    This follows the recommendation of PEP-563 to resolve stringified type
+    annotations at runtime.
+    """
+    all_globals = localns.copy()
+    all_globals.update(localns)
+    for base in inspect.getmro(clazz):
+        all_globals.update(vars(sys.modules[base.__module__]))
+    return all_globals
```
Comment (dosisod) [inline test/mypy_visitor.py:149]: Few things:

* I would use `_class` or `class_` instead of `clazz` (more readable IMO)
* The `all_globals.update(localns)` line shouldn't be needed, since we already are `copy()`ing on the line above

**Review by dosisod — CHANGES_REQUESTED** [review]
Good work so far!

**Comment by jdevera** [comment]
Thank you for the thorough review. I'll give this another spin in the coming days.

File: refurb/visitor/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from .mapping import METHOD_NODE_MAPPINGS
+from .visitor import RefurbVisitor
+
+__all__ = ["METHOD_NODE_MAPPINGS", "RefurbVisitor"]
```
[the remark this answers is not in the document]
Comment (jdevera) [reply refurb/visitor/__init__.py:4]: Happy to make the change, but hear this first :)

I have always used lists because the differences are not significant in terms of performance or readability but there is a somewhat-philosophical argument for lists, it's captured very well here:
https://stackoverflow.com/a/66098991/116957

What do you think? :)

File: refurb/visitor/mapping.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+from typing import Type
+
+import mypy.nodes
+import mypy.patterns
+
+VisitorNodeTypeMap = dict[str, Type[mypy.nodes.Node]]
```
[the remark this answers is not in the document]
Comment (jdevera) [reply refurb/visitor/mapping.py:6]: Ah! This is where Pycharm lied to me! 
CleanShot 2022-10-12 at 18 40 18 - PyCharm - refurb – mapping py@2x

File: test/test_visitor.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+@pytest.fixture
+def dummy_visitor() -> RefurbVisitor:
+    """
+    This fixture provides a RefurbVisitor instance with a visit method for each
+    possible node, but no checks to run.
+
+    This forces method generation but calling the methods does nothing.
+    """
+    checks: Checks = defaultdict(list)
+    checks |= {ty: [] for ty in METHOD_NODE_MAPPINGS.values()}
```
[the remark this answers is not in the document]
Comment (jdevera) [reply test/test_visitor.py:23]: I did try creating from the type alias, but I still had to write the default factory:

```python
checks = Checks(list)
```

That looked weird to me, it creates this question "What is list for" that makes you have to follow `Checks` to see what it is. This is why I went with the `defaultdict(list)` approach with the type annotation, I thought it would answer all questions on the spot.

What I didn't realise is that you could initialise the `defaultdict` with a dictionary as the second parameter.

So may I propose:
```python
    checks = defaultdict(
        list, {ty: [] for ty in METHOD_NODE_MAPPINGS.values()}
    )
    return RefurbVisitor(checks)
```

If you're not convinced, I'll go with your last suggestion, please let me know.

File: test/mypy_visitor.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    @dataclass
+    class PreferPureLoaderHook:
+        orig_hook: Callable[[str], PathEntryFinder]
+
+        def __call__(self, path: str) -> PathEntryFinder:
+            finder = self.orig_hook(path)
+            if isinstance(finder, FileFinder):
+                # Move pure python file loaders to the front
+                finder._loaders.sort(  # type: ignore
+                    key=lambda pair: 0 if pair[0] in (".py", ".pyc") else 1
```
[the remark this answers is not in the document]
Comment (jdevera) [reply test/mypy_visitor.py:69]: Yes, but that relies on the numeric interpretation of True and False, since it is used as a sorting key, and I think the current way, with numbers, is more explicit, even when slightly longer.

File: test/mypy_visitor.py (hunk 3 above)
Comment (jdevera) [reply test/mypy_visitor.py:149]: Thanks, I hated clazz, and this actually made me think of a more specific name instead.

WRT the `update`: 🤦 . Yep.

File: test/conftest.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+pytest_plugins = ["test.mypy_visitor"]
+
+
+@pytest.fixture(scope="session")
+def test_data_path() -> Path:
+    data_path = Path(__file__).parent / "data"
+    assert data_path.exists()
+    assert data_path.is_dir()
+    return data_path
```
[the remark this answers is not in the document]
Comment (jdevera) [reply test/conftest.py:13]: It will be! I meant to use it to solve an issue where tests would fail depending on where I ran them from.

File: test/test_visitor.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        print(method_name)
+        method_types = typing.get_type_hints(method)
+        assert "o" in method_types, f"No 'o' parameter in method {method_name}"
+        node_type = method_types["o"]
+        visitor_mappings[method_name] = node_type
+
+    assert visitor_mappings == METHOD_NODE_MAPPINGS
+
+
+def test_mypy_consistence(mypy_visitor_mapping: VisitorNodeTypeMap) -> None:
```
[the remark this answers is not in the document]
Comment (jdevera) [reply test/test_visitor.py:64]: Started with a session scoped fixture because I had several tests in mind, but then I realised that a simple `assert a == b` would give all the information we need in case of mismatch. I'll change this.

File: refurb/visitor/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dosisod) [reply refurb/visitor/__init__.py:4]: I am fine with it either way. I am just more used to seeing it as a tuple is all. I guess a list would be nice if you plan on dynamically modifying it (like one of the answers mentions), but if you know "these are the only fields I want to export" then a tuple might be better suited.

File: test/test_visitor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dosisod) [reply test/test_visitor.py]: I see you updated this to `checks: Checks = defaultdict(...)`, I presume because `checks = defaultdict(...)` doesn't include a type for `list`, am I correct there?

Either way, I think `checks = Checks(...)` would be more succinct, since you already need to type it as `Checks`, why not just initialize it with `Checks` as well?

File: test/mypy_visitor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dosisod) [reply test/mypy_visitor.py:68]: I see what you mean. I guess in that case you could use `"first"` and `"second"` instead of `0` and `1`, but I think that would be too long!

**Comment by jdevera** [comment]
Latest should have all requested changes.

**Comment by dosisod** [comment]
Everything looks good to me, except for the failing test: In #71 I had to do some dynamic visitor generation, and thus had to import the from the old `_visitor_mappings` file. Should be an easy fix, just need to update the import.

**Comment by jdevera** [comment]
Thank goodness for automated tests, I ran `make all` **before** the last merge conflict 🤦

**Review by dosisod — APPROVED** [review]
Great! Looks good to me!

**Comment by jdevera** [comment]
@dosisod I'm taking part in [Hacktoberfest](https://hacktoberfest.com/) this year. If you found this PR valuable, can ask you to please label it as `hacktoberfest-accepted` so it counts towards my monthly PR count?

**Comment by dosisod** [comment]
Done!
